### ISSUE TITLE
`generic_const_exprs` does not work properly with const/type param defaults

### ISSUE THREAD

**Issue by BoxyUwU** [rust-lang/rust#106994]
In order to make `feature(generic_const_exprs)` not completely broken when used with `feature(const_generics_defaults)`, #86580 was landed. From the description of the PR:

> This PR doesn't handle the predicates of the const so
> ```rust
> trait Foo<const N: usize> { const Assoc: usize; }
> pub struct Bar<const N: usize = { <()>::Assoc }> where (): Foo<N>;
> ```
> Resolves to `<() as Foo<N>>::Assoc` which can allow for using fwd declared params indirectly.
> 
> ```rust
> trait Foo<const N: usize> {}
> struct Bar<const N: usize = { 2 + 3 }> where (): Foo<N>;
> ```
> This code also ICEs under this PR because instantiating the default's predicates causes an ICE as predicates_of contains predicates with fwd declared params

because of this these two features are still pretty incompatible with eachother. This issue tracks this as there have been a large volume of issues that are all about which makes looking through `F-generic_const-exprs` issues harder than it needs to be.

It was previously attempted to fix this in #106847 but that was closed:

> I don't feel super comfortable merging more hacks to make `generic_const_exprs` + `const_generics_defaults` work in the presence of each other. #86580 alone was already rather hacky and I think this PR just makes it way too much to be reasonable.  It's also unclear to me whether this is going to interfere with fixing some of the other issues we have with const generics and I don't want to make that any harder than necessary. 
> 
> `feature(generic_const_exprs)` is an incomplete feature (both literally and according to the `incomplete_features` lint), having it in a broken state because we are not currently in a good position to properly fix this seems fine to me. Thanks for making the PR anyway but I expect this issue will stay open until we make a lot more progress on `generic_const_exprs` :sweat_smile:  
    

### duplicate issues

When this is fixed the following should be revisited and checked to make sure everything works as intended:

- [ ] [#106473 description](https://github.com/rust-lang/rust/issues/106473#issue-1519743404)
- [ ] [#106473 comment](https://github.com/rust-lang/rust/issues/106473#issuecomment-1373778123)
- [ ] [#101036 comment](https://github.com/rust-lang/rust/issues/101036#issuecomment-1228413898)
- [ ] [#101036 comment 2](https://github.com/rust-lang/rust/issues/101036#issuecomment-1345297071)
- [ ] [#94846 description](https://github.com/rust-lang/rust/issues/94846#issue-1166441076)
- [ ] [#93182 description](https://github.com/rust-lang/rust/issues/93182#issue-1111159495)
- [ ] [#105631 description](https://github.com/rust-lang/rust/issues/105631#issue-1492669743)
- [ ] [#111055 description](https://github.com/rust-lang/rust/issues/111055#issue-1690612999)
- [ ] [#111419 description](https://github.com/rust-lang/rust/issues/111419#issue-1703197021)
- [ ] [#117496 description](https://github.com/rust-lang/rust/issues/117496#issue-1973053401)

**Comment by peter-kehl** [comment]
Thank you for identifying this. And for any progress (hopefully).
Weeks of my proof-of-concept work on co-operative allocation in `library/alloc` depend on this. (I've referenced the old/duplicate of this issue in my commits; I'll reference this issue from new commits and rebase.)

Minimizing this even more - even without any declared trait:
```
#![allow(incomplete_features)]
#![feature(generic_const_exprs)]

struct V<const U: usize = {1+2}>
where
    [(); U]:;
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=971190f4162846600db09b2f4282f754

**Comment by peter-kehl** [comment]
Hoping to help narrow this closer: This fails also if the `const` expression has no arithmetics, just type casting:
```rust
#![allow(incomplete_features)]
#![feature(generic_const_exprs)]

struct V<const U: usize = {0i32 as usize}>
where
    [(); U]:;
```

**Comment by peter-kehl** [comment]
What other combinations are worth checking? So that I can help this forward, please?

**Comment by BoxyUwU** [comment]
I do not believe there is anything worth checking, this issue is relatively well understood already. The problem is how to solve it, not figuring out what the cause is

**Comment by peter-kehl** [comment]
Still, while I'm narrowing this down from the "consumer's" point of view, and searching for workarounds, here are 2 examples.
1. ICE, but ONLY if this generic type is actually used (in a function signature, an expression...). Otherwise the type definition itself does compile.
```rust
#![allow(incomplete_features)]
#![feature(generic_const_exprs)]

trait Tr {
    const C: usize = 0;
}

struct Str {}
impl Tr for Str {}

// ICE, but ONLY if this generic type is used (in a function signature, an expression...).
struct ICE<T: Tr = Str, const C: usize = {T::C}>
where [(); C]:,
{
    _t: core::marker::PhantomData<T>,
    _arr: [(); C],
}

// Using the above type in a function signature, or in an expression causes an ICE.

fn _type_in_fn_sig_return_causes_ice() -> ICE {
    loop{}
}
// and/or:

fn _type_in_fn_sig_param_causes_ice(_: &ICE) {
    loop{}
}
// and/or:

fn _instantiate() {
    ICE::<Str> {
        _t: core::marker::PhantomData {},
        _arr: []
    };
}
```
2. ICE caused by mere presence of `#![feature(generic_const_exprs)]` itself. This code doesn't actually need `generic_const_exprs`. But, if the feature is enabled, then this code has an ICE.
```rust
#![allow(incomplete_features)]
// `struct ICE` below compiles well WITHOUT generic_const_exprs. But it does have an ICE with
// generic_const_exprs!
//
// The ICE is triggered even if the const generic-based struct itself (`ICE` below) is not used at
// all (not in any function signature, expression...). (And regardless of whether the type is
// public.)
#![feature(generic_const_exprs)]

trait Tr {
    const C: usize = 0;
}

struct Str {}
impl Tr for Str {}

struct ICE<const C: usize={Str::C}>
where [(); C]:,
{
    _arr: [(); C],
}
```
